Make Preferences available in the back end

tsmaeder opened the pull request:
#### What it does

The PR makes preferences support available in the back end. Only default and user preferences can be accessed in the back end. The API has changed in the following ways:
- Many files have been moved from the `browser` folder to the `common` folder. Imports will have to be adapted
- `PreferenceSchemaProvider` has been replaced by two separate `PreferenceSchemaServiceImpl` (and corresponding interface) and `DefaultsPreferenceProvider` classes.
- Preference schema typing has been simplified: a preference schema is no longer extending IJSONSchema and typing has been adapted to strictly use Theia types (for example for scopes) and a straight-forward extension of standard IJSONSchema for properties. This means schemas from VS Code (contributed via extensions) must be converted to Theia format.
- `PrefenceSchemaService` separates between adding a schema and registering a default override for a property. Also, the service uses explicit override identifiers instead of encoding the override in the preference key. The schema service strictly distinguishes between preference schema and the derived JSON Schema for preference files. `JSONValue` is now used instead of `any` where applicable. Schema properties must be added before overrides are registered.
`PreferenceSchemaService` now has the concept of`validScopes`. In the back end, only`Default` and `User` can be used.
- `PreferenceContribution` now has a `initSchema()` method in addition to the declarative Schema contribution. It is used to register overrides.  

Fixes https://github.com/eclipse-theia/theia/issues/15635

#### How to test

The PR adds three commands: "Get Backend Preference", "Inspect Backend Preference" and "Set Backend Preference". To test front-end prefernces, use the preference UI.

#### Follow-ups
https://github.com/eclipse-theia/theia/issues/16046
https://github.com/eclipse-theia/theia/issues/16044

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

#### Breaking changes

- [x] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Attribution
Contributed on behalf of STMicroelectronics

#### Review checklist

- [ ] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

sdirix commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -63,13 +64,13 @@ export class TaskConfigurationManager {
 
     @inject(PreferenceProvider) @named(PreferenceScope.Folder)
-    protected readonly folderPreferences: PreferenceProvider;
+    protected readonly folderPreferences: PreferenceProviderImpl;
```
Why do we use the `Impl` type? Isn't `PreferenceProvider` sufficient?

sdirix commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -20,18 +20,19 @@ import { inject, injectable, postConstruct, named } from '@theia/core/shared/inv
 import { FileService } from '@theia/filesystem/lib/browser/file-service';
-import { DisposableCollection } from '@theia/core/lib/common';
 import { TaskSchemaUpdater } from './task-schema-updater';
+import { JSONObject } from '@theia/core/shared/@lumino/coreutils';
```
Again import from Lumino

tsmaeder replied:
Again: auto-import.

sdirix commented on packages/task/src/browser/task-configuration-model.ts:
```diff
@@ -18,8 +18,9 @@ import URI from '@theia/core/lib/common/uri';
-import { isObject } from '@theia/core/lib/common';
+import { PreferenceProviderDataChanges, PreferenceProviderDataChange, isObject } from '@theia/core/lib/common';
+import { PreferenceProvider } from '@theia/core/lib/common/preferences/preference-provider';
+import { JSONValue } from '@theia/core/shared/@lumino/coreutils';
```
Lumino import

tsmaeder replied:
Where do we have JSON types in @theia/core?

sdirix commented on packages/task/src/node/task-server.slow-spec.ts:
```diff
@@ -424,6 +429,7 @@ function checkSuccessfulProcessExit(taskInfo: TaskInfo, taskWatcher: TaskWatcher
 function getExitStatus(taskInfo: TaskInfo, taskWatcher: TaskWatcher): Promise<string | number> {
     return new Promise<string | number>((resolve, reject) => {
         taskWatcher.onTaskExit((event: TaskExitedEvent) => {
+            console.log(`task ${event.taskId} exited, ${JSON.stringify(event)}`);
```
Do we want to log this on `log` level or rather `debug`?

sdirix commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -70,11 +71,11 @@ export class WorkspaceService implements FrontendApplicationContribution, Worksp
+    @inject(PreferenceService)
+    protected readonly preferenceImpl: PreferenceService;
 
-    @inject(PreferenceSchemaProvider)
-    protected readonly schemaProvider: PreferenceSchemaProvider;
+    @inject(PreferenceSchemaService)
+    protected readonly schemaProvider: PreferenceSchemaService;
```
We should rename the variables

sdirix commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -603,10 +604,7 @@ export class WorkspaceService implements FrontendApplicationContribution, Worksp
-                if (this.schemaProvider.isValidInScope(p, PreferenceScope.Folder)) {
-                    continue;
-                }
+            for (const p of Object.keys(this.schemaProvider.getJSONSchema(PreferenceScope.Workspace).properties!)) {
```
We now switch here from `PreferenceScope.Folder` to `PreferenceScope.Workspace`. Is this correct?

tsmaeder replied:
We used to filter out the preferences that could possibly be overridden in folder scope from being saved in the workspace file (if I read the code correctly). I don't think this makes sense: if someone makes a setting in a multi-folder workspace, it should be saved into the workspace file, no?

sdirix reviewed (commented):
Thanks for the great implementation. The whole preferences code definitely feels cleaner now!

I did some basic testing (the api-sample commands are great!) with workspace and user settings which worked great for me.

I left some (94 :sweat_smile:) comments and questions. Please have a look

tsmaeder commented on examples/api-samples/src/node/api-samples-backend-module.ts:
```diff
@@ -21,8 +21,21 @@ import { SampleMockOpenVsxServer } from './sample-mock-open-vsx-server';
+import { SampleBackendPreferencesBackendServiceImpl } from './sample-backend-preferences-service';
+import { ConnectionContainerModule } from '@theia/core/lib/node/messaging/connection-container-module';
+
+const samplesConnectionContainerModule = ConnectionContainerModule.create(({ bind, bindBackendService }) => {
+    bind(SampleBackendPreferencesBackendServiceImpl).toSelf().inSingletonScope();
+    bind(SampleBackendPreferencesService).toService(SampleBackendPreferencesBackendServiceImpl);
+    bindBackendService(sampleBackendPreferencesServicePath, SampleBackendPreferencesService);
+});
```
How else would you bind a back end service that talks to the front end?

tsmaeder commented on packages/core/src/browser/frontend-application-bindings.ts:
```diff
@@ -19,25 +19,27 @@ import {
 }
 
 export function bindPreferenceService(bind: interfaces.Bind): void {
-    bind(PreferenceProvider).toSelf().inSingletonScope().whenTargetNamed(PreferenceScope.User);
-    bind(PreferenceProvider).toSelf().inSingletonScope().whenTargetNamed(PreferenceScope.Workspace);
-    bind(PreferenceProvider).toSelf().inSingletonScope().whenTargetNamed(PreferenceScope.Folder);
```
I've become convinced over the years that the practice of binding dummy implementations for the convenience of adopters is a mistake. It forces us to do a rebind in our own code and if someone wants to override the service with their own version, they'll be at the mercy of the order of initialization of the theia modules. That order is not defined. IMO, we should never rely on "rebind" in our own code base, because rebind is a coin you can only spend once.
I realize this is a general discussion, though.

as-salam replied:
> I've become convinced over the years that the practice of binding dummy implementations for the convenience of adopters is a mistake. It forces us to do a rebind in our own code and if someone wants to override the service with their own version, they'll be at the mercy of the order of initialization of the theia modules. That order is not defined. IMO, we should never rely on "rebind" in our own code base, because rebind is a coin you can only spend once.
> I realize this is a general discussion, though.
> 

I agree, in Pandino such dummies are irrelevant because there's the possibility to tell the SCR which service is prioritized by its level and also the fact that all is declarative and managed automatically. It saves time greatly.

Binding dummies is a dumb idea.

Cheers

planger replied:
The solution @tsmaeder proposes seems sound to me, even though it requires adopters to most likely have to do code changes. But it is properly mentioned in the migration guide. So I guess this is fine.

tsmaeder commented on packages/core/src/browser/preferences/frontend-config-preference-contributions.ts:
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable @typescript-eslint/quotes */
+// *****************************************************************************
+// Copyright (C) 2018 Ericsson and others.
```
It used to be part of `preferences-contribution.ts` and the code mainly comes from there, so I think the copyright notice is right, here.

tsmaeder commented on packages/core/src/browser/preferences/preference-validation-service.ts:
```diff
@@ -352,7 +350,29 @@ export class PreferenceValidationService {
+        if (property.default !== undefined) {
+            return property.default;
+        }
+        const type = Array.isArray(property.type) ? property.type[0] : property.type;
```
I copied this verbatim from `preferences-contribution.ts` and would rather keep it just like before.

tsmaeder commented on packages/core/src/common/core-preferences.ts:
```diff
@@ -229,14 +227,14 @@ export const corePreferenceSchema: PreferenceSchema = {
             enum: ['dark', 'light', 'hc-theia'],
             enumItemLabels: ['Dark (Theia)', 'Light (Theia)', 'High Contrast (Theia)'],
-            default: DefaultTheme.defaultForOSTheme(FrontendApplicationConfigProvider.get().defaultTheme),
+            default: 'light',
```
No, the front end registers a default override for this property. You cannot determine the defautl theme for the OS if you don't know the OS of the front end.

tsmaeder commented on packages/core/src/common/core-preferences.ts:
```diff
@@ -229,14 +227,14 @@ export const corePreferenceSchema: PreferenceSchema = {
             enum: ['none', 'theia-file-icons'],
             enumItemLabels: [nls.localizeByDefault('None'), 'File Icons (Theia)'],
-            default: FrontendApplicationConfigProvider.get().defaultIconTheme,
+            default: 'none',
```
yes

tsmaeder commented on packages/core/src/common/preferences/defaults-preference-provider.ts:
```diff
@@ -0,0 +1,91 @@
+                    changes[change.preferenceName] = change;
+                }
+            }
+            this.emitPreferencesChangedEvent(changes);
```
No, in that case, the preference schema service should not have emitted a change. Also, default changes only happen at startup (when no-one should be listening yet) and when plugins are loaded. I don't think it's a problem.

tsmaeder commented on packages/core/src/common/preferences/index.ts:
```diff
@@ -0,0 +1,27 @@
+// *****************************************************************************
+// Copyright (C) 2017 TypeFox and others.
```
I copied and changed the file from the `browser` folder.

sdirix replied:
It's a new index file. All index files in Theia contain links relative to the folder, i.e. there is no "creativity" in index files in the Theia project. I don't see how the copyright notice of a similar file would influence this one, even if the other file was used as a template to save effort. Keeping the Typefox 2017 copyright notice here would mean all index files in the Theia project should have Typefox 2017 copyright notice.

On the other side, the index file which was copied, i.e. the `packages/core/src/browser/preferences/index.ts` should keep its old copyright notice. In this PR the other copyright notice was changed which we don't do in the Theia project.

tsmaeder commented on packages/core/src/common/preferences/preference-proxy.ts:
```diff
@@ -193,10 +194,10 @@ export function createPreferenceProxy<T>(preferences: PreferenceService, promise
                 const overridden = preferences.overriddenPreferenceName(e.preferenceName);
                 const preferenceName: any = overridden ? overridden.preferenceName : e.preferenceName;
```
Unfortunately, that does not compile, as `keyof T !== 'string'`

sdirix commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+            };
+        }
+        const promises: Promise<void>[] = [];
+        // this.readConfiguredPreferences(); => needs separate contribution frontend/back end
```
Then let's remove it :)

tsmaeder commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+            // Update the property with new values
+            const updatedProperty = { ...existing, ...property };
+            this.properties.set(key, updatedProperty);
+            if (this.defaultOverrides.get(key)?.get(NO_OVERRIDE) === undefined && property.default !== existing.default) {
```
The worst thing that could happen is that we send a spurious change notification for an edge case. Most of the time, the values will not be the same, so the performance hit of the deepEqual might be worse.

tsmaeder replied:
I think `deepEqual()` would be correct.

tsmaeder commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+            }
+            // handle case where old property was not overrideable and vice versa
+
+            this.setJSONSchemasProperty(key, updatedProperty);
+            this.schemaChangedEmitter.fire();
```
Not sure the optimisation is worth the extra complexity: 99% of the time, the property _will_ have changed. Otherwise, why invoke the method?

tsmaeder commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+        if (overrides) {
+            const values = overrides.get(overrideId);
+            if (values) {
+                return values[0][1]; // there will be no empty values arrays in the data structure
```
You are correct that this edge case is not handled. The right fix would be to use `unshift` instead of `push`: we're reading default overrides much more frequently than writing them.

tsmaeder replied:
I've mad the change to `unshift`

tsmaeder commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+        if (overrides) {
+            const values = overrides.get(overrideId);
+            if (values) {
+                return values[0][1]; // there will be no empty values arrays in the data structure
```
see above

tsmaeder commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+                if (overrides) {
+                    for (const [overrideId, values] of overrides.entries()) {
+                        if (overrideId === NO_OVERRIDE) {
+                            result[key] = values[0][1];
```
see above

tsmaeder commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+                        } else {
+                            const overrideKey = `[${overrideId}]`;
+                            const target: JSONObject = result[overrideKey] as JSONObject || {};
+                            target[key] = values[0][1];
```
see above

tsmaeder commented on packages/core/src/common/preferences/preference-schema.ts:
```diff
@@ -16,89 +16,162 @@
 
-export interface PreferenceDataSchema {
-    [name: string]: any,
+export interface PreferenceSchema {
```
I've thought about that, but I think in the long run and for new adopters it's easier to work with correct typing. Mixing IJSsonSchema and the preference schema was one of the confusing aspects of the previous schema code. It's a trivial change to make that is caught by the compiler. In general, I don't expect adopters to contribute very many preference schemas.

tsmaeder commented on packages/core/src/common/preferences/preference-schema.ts:
```diff
@@ -16,89 +16,162 @@
-    patternProperties: {
-        [name: string]: PreferenceDataProperty
-    };
+    title?: string,
+    defaultOverridable?: boolean;
+    properties: IndexedAccess<PreferenceDataProperty>;
```
It wasn't before, but you're right.

tsmaeder commented on packages/core/src/common/preferences/preference-schema.ts:
```diff
@@ -16,89 +16,162 @@
+
+export interface DefaultValueChangedEvent {
+    key: string;
+    overrideIdentifier?: string;
+    otherAffectedOverrides: string[];
```
No, when the default for a base property changes, those overrides that do not have their own default will be affected and will be reported here.

tsmaeder commented on packages/core/src/common/preferences/preference-scope.ts:
```diff
@@ -14,6 +14,11 @@
 // *****************************************************************************
 
+/**
+ * An array of preference scopes that are valide in a given context, sorted from more general to more specific
```
While that may be the current use case, there is nothing preventing someone from binding a different preference service inside a connection container, for example.

tsmaeder commented on packages/core/src/common/preferences/preference-service.ts:
```diff
@@ -338,30 +337,32 @@ export class PreferenceServiceImpl implements PreferenceService {
                 }
             }
-            if (this.schema.isValidInScope(preferenceName, PreferenceScope.Folder)) {
+            if (this.schema.isValidInScope(baseName, PreferenceScope.Folder)) {
```
I don't understand the question? Anyway, I did not really change the algorithm here.

tsmaeder replied:
I've implemented https://github.com/eclipse-theia/theia/issues/16176 in this PR.

tsmaeder commented on packages/git/src/browser/git-scm-provider.spec.ts:
```diff
@@ -39,7 +39,7 @@ import { DefaultGitEnvProvider, GitEnvProvider } from '../node/env/git-env-provi
 import { bindGit } from '../node/git-backend-module';
```
I didn't make it up 🤷

tsmaeder commented on packages/plugin-ext/src/hosted/node/scanners/scanner-theia.ts:
```diff
@@ -191,6 +238,21 @@ export class TheiaPluginScanner extends AbstractPluginScanner {
 
+    static getScope(monacoScope: string | undefined): { scope: PreferenceScope | undefined, overridable: boolean } {
+        switch (monacoScope) {
+            case 'machine-overridable':
```
The scope "machine-overridable" does not really map to anything in Theia. It did not before, either: the `fromString()` method would simply return undefined, thus falling back to the default scope of the schema.

tsmaeder replied:
Actually, I think it's the correct mapping to Theia concepts if you look at the doc in https://code.visualstudio.com/api/references/contribution-points#contributes.configuration

sdirix commented on packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts:
```diff
@@ -53,6 +51,7 @@ import { PluginTerminalRegistry } from './plugin-terminal-registry';
 import { ContextKeyService } from '@theia/core/lib/browser/context-key-service';
 import { LanguageService } from '@theia/core/lib/browser/language-service';
 import { ThemeIcon } from '@theia/monaco-editor-core/esm/vs/base/common/themables';
+import { JSONObject, JSONValue } from '@theia/core/shared/@lumino/coreutils';
```
I didn't realize that the Lumino types already sneaked into our code base. I thought this came in accidentally via this PR.

You don't need to change it here but it's crazy to me to introduce dependencies to 3rd party frameworks for such simple types. They should be defined in our core package instead.

That's all it is:

```
export type JSONPrimitive = boolean | number | string | null;

export type JSONValue = JSONPrimitive | JSONObject | JSONArray;

export interface JSONObject {
  [key: string]: JSONValue;
}

export interface JSONArray extends Array<JSONValue> {}
```

tsmaeder replied:
No, we're also using `deepEquals()` for example. In total, the module is about 400 loc.

sdirix commented on packages/preferences/src/browser/monaco-jsonc-editor.ts:
```diff
@@ -33,9 +33,9 @@ export class MonacoJSONCEditor {
     getEditOperations(model: MonacoEditorModel, path: jsoncparser.JSONPath, value: unknown): monaco.editor.IIdentifiedSingleEditOperation[] {
         const textModel = model.textEditorModel;
-        const content = model.getText().trim();
+        const content = model.getText();
```
Makes sense. Can this explain why Theia sometimes corrupted the settings.json?

tsmaeder commented on packages/preferences/src/browser/util/preference-tree-generator.ts:
```diff
@@ -88,12 +86,12 @@ export class PreferenceTreeGenerator {
-            if (preference in preferencesSchema.properties) {
-                this.createLeafNode(preference, commonlyUsed, preferencesSchema.properties[preference]);
+            if (preference in properties) {
+                this.createLeafNode(preference, commonlyUsed, properties.get(preference)!);
```
No, here' we're iterating over the settings in `commonlyUsedLayout`, not all the properties in the schema

tsmaeder commented on packages/preferences/src/common/abstract-resource-preference-provider.ts:
```diff
@@ -20,21 +20,52 @@
-import { FileService } from '@theia/filesystem/lib/browser/file-service';
-import { PreferenceContext, PreferenceTransaction, PreferenceTransactionFactory } from './preference-transaction-manager';
 import { Emitter, Event } from '@theia/core';
+import { JSONValue } from '@theia/core/shared/@lumino/coreutils';
```
I assume by your earlier comment that you're realized we use the JSONValue support from @lumino. There is nothing wrong with that, as it works on node.

tsmaeder commented on packages/core/src/common/preferences/preference-schema-service.ts:
```diff
@@ -0,0 +1,428 @@
+            if (this.defaultOverrides.get(key)?.get(NO_OVERRIDE) === undefined && property.default !== existing.default) {
+                this.defaultValueChangedEmitter.fire(this.changeFor(key, undefined, this.defaultOverrides.get(key), undefined, property.default!));
+            }
+            // handle case where old property was not overrideable and vice versa
```
That was a comment to myself whether we need to consider whether the property has a changed `overridable` flag. But since we're not checking the `overridable` flag when we register an override value, it does not make sense to check for it here.

tsmaeder commented on packages/core/src/browser/preferences/frontend-config-preference-contributions.ts:
```diff
@@ -0,0 +1,53 @@
+        if (FrontendApplicationPreferenceConfig.is(config)) {
+            service.registerOverride('workbench.colorTheme', undefined, DefaultTheme.defaultForOSTheme(config.defaultTheme));
+            if (config.defaultIconTheme) {
+                service.registerOverride('workbench.iconTheme', undefined, config.defaultIconTheme);
```
`ApplicationConfig.defaultIconTheme` is not the same as the `workbenche.iconThema` preference, hence the difference in names. I believe there is a case to be made to keep the core preference as much as possible the same between front-end and back-end. This way, the typing, documentation, etc. don't diverge.

tsmaeder commented on packages/getting-started/src/common/getting-started-preferences.ts:
```diff
@@ -17,24 +17,22 @@
                 nls.localizeByDefault('Start without an editor.'),
                 nls.localize('theia/getting-started/startup-editor/welcomePage', 'Open the Welcome page, with content to aid in getting started with {0} and extensions.',
-                    FrontendApplicationConfigProvider.get().applicationName),
+                    'Theia'),
```
The background is that localization does not work on the back end: you can invoke it, but it will always return the English text.

tsmaeder replied:
The getting started widget is _not_ a preference contribution 🤷

tsmaeder commented on packages/plugin-ext/src/main/browser/preference-registry-main.ts:
```diff
@@ -42,7 +42,7 @@ export function getPreferences(preferenceProviderProvider: PreferenceProviderPro
         if (scope === PreferenceScope.Folder) {
             for (const f of folders) {
-                const folderPrefs = provider.getPreferences(f);
+                const folderPrefs = provider.getPreferences();
```
No, good catch.

tsmaeder commented on packages/getting-started/src/browser/getting-started-contribution.ts:
```diff
@@ -66,6 +70,25 @@ export class GettingStartedContribution extends AbstractViewContribution<Getting
         });
     }
 
+    async initSchema(service: PreferenceSchemaService): Promise<void> {
```
Which "PreferenceContribution for the frontend"? This _is_ the front end contribution for the "getting-started" module.

tsmaeder commented on packages/core/src/node/backend-application-module.ts:
```diff
@@ -140,4 +155,23 @@ export const backendApplicationModule = new ContainerModule(bind => {
+    bindPreferenceConfigurations(bind);
+    bind(ValidPreferenceScopes).toConstantValue([PreferenceScope.Default, PreferenceScope.User]);
+    bindContributionProvider(bind, PreferenceContribution);
+    bind(PreferenceProviderProvider).toFactory(ctx => (scope: PreferenceScope) => {
```
and ctx.getNamed().... should not throw an exception, right? Just return undefined?

tsmaeder commented:
@sdirix I have addressed all your comments and resolved the ones I could resolve unilaterally. Please have a look at the remaining unresolved comments and close if satisfied.

tsmaeder commented:
@msujew now would be a good time for a review as I won't have time next month ;-)

ndoschek reviewed (commented):
Hi @tsmaeder,  
Thanks for this great improvement! I tested it, and it looks great to me.  
I started looking into the detailed changes, and so far everything seems fine.  
I'll continue on Monday though.
I also asked Philip to take a high-level look as well.  
Thanks again!

planger commented on packages/ai-anthropic/src/node/anthropic-backend-module.ts:
```diff
@@ -16,12 +16,14 @@
 
 // We use a connection module to handle AI services separately for each frontend.
 const anthropicConnectionModule = ConnectionContainerModule.create(({ bind, bindBackendService, bindFrontendService }) => {
+
```
```suggestion
```

planger commented on examples/api-tests/src/preferences.spec.js:
```diff
@@ -19,27 +19,29 @@
-    /** @type PreferenceSchemaProvider */
-    const schemaProvider = container.get(PreferenceSchemaProvider);
+    /** @type PreferenceSchemaService */
+    const schemaProvider = container.get(PreferenceSchemaService);
```
This comment seemed to have slipped through the cracks.

planger reviewed (approved):
Thanks for this substantial change! I didn't thoroughly test it and only looked through the changes on a high level and it seems pretty good to me, even though it requires potentially quite some changes in adopters' code. But it seems properly documented on the migration guide.

ndoschek commented on packages/ai-chat/src/node/ai-chat-backend-module.ts:
```diff
@@ -0,0 +1,25 @@
+// *****************************************************************************
+// Copyright (C) 202 STMicroelectronics and others
```
```suggestion
// Copyright (C) 2025 STMicroelectronics and others
```

ndoschek commented on packages/ai-code-completion/src/node/ai-code-completion-backend-module.ts:
```diff
@@ -0,0 +1,23 @@
+// *****************************************************************************
+// Copyright (C) 2025 Lonti.com Pty Ltd.
```
nit: as the commit is contributed on behalf of ST, is this header correct?

ndoschek reviewed (commented):
Thanks @tsmaeder, it still looks great to me. I had a closer look at the code and only added a small nitpick inline.

ndoschek reviewed (approved):
Thanks!
